### PR TITLE
Improve accessibility of the templates list Actions column

### LINKED ISSUE #42505: Improve the accessibility of the Templates browser 'Actions' column
### Description

As a low vision user, I struggle to see the ellipsis in the 'Actions' column of the Templates browser.

At first, I didn't notice there were some buttons in this column. Even now that I know there are buttons, the ellipsis icon is so tiny that I struggle to see it and click reliably on it.

Besides the very light icon, there are other factors that make these buttons hard to see for low vision users:
- The third column header text is visually hidden. Actually, there's an 'Actions' text rendered within a `VisuallyHidden` component. I guess that's because the buttons in the column cells are displayed conditionally so that sometimes the cells are empty. 
-  The buttons are displayed conditionally. Under some conditions, there are no buttons at all so that it's hard to discover the third column _might_ actually have some content.
- The ellipsis button opens a dropdown menu, even when the menu contains just one item. In this case, users are forced to  click twice to actually take the _only_ available action. Seems to me this isn't ideal. When there's only one action, there's no need for a dropdown menu in the first place. Placing the only available action control directly within the table cell would probably be a better option.

Proposal:
- Always display the column header 'Actions'.
- When the action cell is empty, display some text to clarify there are no actions e.g. 'None'.
- Reconsider the usage of the dropdown menu and use buttons within the cell instead.
- Alternatively, keep the icon button but use an icon with a more visible shape.

### Step-by-step reproduction instructions

- Go to the Site editor.
- Create a Taxonomy template.
- Create a Custom template.
- Make sure you have a default template with some saved customizations.
- Click on 'Toggle navigation' to open the Navigation left sidebar.
- Click on 'Templates to open the templates browser.
- Observe the content of the table third column.
- Observe the third column header text is visually hidden.
- Observe your customized templates have an ellipsis icon button in the third column.
- Explore the content of the ellipsis dropdown menu.

### Screenshots, screen recording, code snippet

The ellipsis icon button are barely visible for low vision users:

Screenshot 2022-07-18 at 15 06 18

Dropdown menu with only one available action:

Screenshot 2022-07-15 at 10 55 54

One more case where the dropdown menu contains only one action:

Screenshot 2022-07-15 at 12 30 33

Dropdown menu with two actions:

Screenshot 2022-07-15 at 10 56 00

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by afercia**
Fixes https://github.com/WordPress/gutenberg/issues/42505

## What?

Changes the Actions button in the templates list to use text instead of only an icon. Also makes the column header visible to as there's no reason to hide it.

## Why?

The ellipsis icon of the Actions button is barely visible for low vision users. In other places, for example in the block toolbar, the ellipsis icon is more visible because of the context and the borders. Instead, in the templates list is very light.

From an accessibility perspective, the golden rule for 'icon-only' buttons should be: use them only when there's really no space to use visible text. In this case, there's enough space to use tedxt and there's no reason to use only an icon.

## How?

- Changes the dropdown menu toggle button to use visible text + a chevron icon instead of only the ellipsis icon.
- Makes the column header text visible and changes it from 'Actions' to 'Manage' to avoid repetition.
- **Removes the flex styles used for the column**.  This is necessary to make the column header text and the buttons within the cells align correctly. I'm not sure why flexbox was used in the first place.
- Since this is a _data_ table, make sure to not have empty cells and use visually hidden text for the 'no actions' case.

## Testing Instructions
- Go to the Site Editor.
- Make sure to have a default template with some customization and a custom template.
- Click Toggle navigation, and then click Templates.
- Observe the third column header text is now visible.
- Observe the customized templates have a button in the third column. The button text is 'Actions' plus a chevron icon.
- Observe the templates with no actions available have an icon representing a line in the third column and that there's a visually hidden text 'No actions available'. This pattern is consistent with the one already used in the WordPress list tables.
- Check there are no breaking style changes.
- Worth noting this table was not fully responsive already. The CSS changes don't improve responsiveness but seems to me they don't make it worse either.

## Screenshots or screencast 

Before:

before

After: 

after

### CODE REVIEW

**Comment by jameskoster** [comment]
I think there's a general usability consideration here, additional the a11y issues. Ellipsis icon buttons that open an option / action menu is a fairly well established convention in the Editor. It appears in the document toolbar, block toolbar, list view, global styles, design tools, etc. It might be good to explore a solution that fits all uses these cases and maintains holistic consistency / predictability.

One option could be to have these buttons respect the 'Show button text labels' option in the Editor, so that they switch to text when appropriate? I appreciate that doesn't necessarily help low-vision users, but perhaps there are other approaches to look at?

I noticed in the PR that the horizontal stroke between rows has disappeared, which doesn't seem to be intentional?

cc @javierarce since I know you've been looking at control designs for https://github.com/WordPress/gutenberg/issues/39082.

**Comment by afercia** [comment]
@jameskoster thanks for your feedback. I'm not sure this is a case where the 'Show button text labels' option would fit. The 'icon-only' buttons issue has been discussed at length and I do realize is a bit controversial.  As general accessibility (and I's say also usability) principle, visible text is preferable because it works for all users.  'icon-only' buttons should only be used when there's not enough space for text.


**TL:DR**
A few more considerations:

> Ellipsis icon buttons that open an option / action menu is a fairly well established convention in the Editor. It appears in the document toolbar, block toolbar, list view, global styles, design tools, etc.

I think the ellipsis icon buttons are sometimes used in a slightly inconsistent way in the Editor. To me, the 'ellipsis' means 'More stuff', that is: a menu with (less often used) controls _in addition_ to a set of other controls. The concept of 'more' makes perfectly sense in the context of a toolbar, where there's a set of controls displayed by default and then there's a 'More' button that groups other controls in a menu. This applies for example to the document toolbar, the block toolbar and the like.

Instead, to me, the ellipsis doesn't make much sense when used 'in isolation', as there are no other controls where the ellipsis button is supposed to add 'more' things to. This applies, for example, to the list view, global styles, and the template details popover, where the ellipsis just opens a menu with options / actions. In that context, the button has a more generic meaning on 'Options / Actions' and maybe it should use a different icon.

In fact, in other places of the UI, options / actions menus use other icons (the cog icon, for example).

I'd think a more important usability issue, as mentioned in https://github.com/WordPress/gutenberg/issues/42505, is that the dropdown menu is used even when it contains only one item:

only one item

In this case, a dropdown menu doesn't seem appropriate in the first place, because it forces users to unnecessarily click twice. That's not a great experience for users. When there's only one action, there should be just a button for that action, not a dropdown. This consideration is unrelated to this PR though and I'll create a separate issue, but it does have an impact on the ellipsis button.

Lastly, I think the pattern 'text + chevron icon' is not new in the Editor and it's probably more usable than the ellipsis button.

**Comment by javierarce** [comment]
I agree that using an ellipsis is difficult to justify in this case, and a dropdown with just one item is frustrating.

If we just have two actions (and not for every case) could we consider a text-based approach? It's less scalable, but it could work for the time being. 

image

In this case, we wouldn't need to indicate 'No actions available', because the distinction would be clear. 

Another thing we could do in addition to this change is to group the templates into editable and not editable ones. It would probably require explaining the distinction, but here's a quick example of what I mean:

image

**Comment by jameskoster** [comment]
We're not far away from suggesting the Post Actions UI from wp-admin at this point :D 

Screenshot 2022-07-21 at 18 01 08

One potential issue with just listing the links however, is accommodating the 'Clear customizations' button which is fairly long and includes a description:

Screenshot 2022-07-21 at 18 03 15

It could be worth exploring how all the combinations of links fit in, and how everything stacks up on small screens.

---

I think there could be something to the idea of separating the different template types (unedited, customized, user-generated), but wonder if that's something to do via filtering rather an a permanent layout configuration 🤔

**Comment by jameskoster** [comment]
I appreciate this goes way beyond the scope of this issue, but perhaps a card-style layout with previews could work better than the current table?

Screenshot 2022-07-21 at 19 15 41

It's easy to imagine the number of columns expanding / contracting according to screen size.

**Comment by paaljoachim** [comment]
@jameskoster That is interesting! It could perhaps show a thumbnail of what the template looks like! Card views really gives a much better overview of each template. An approach that can be further explored!

**Comment by afercia** [comment]
Thanks all for your feedback.

While a table would be semantically more appropriate to represent the templates data, I'd agree it's worth exploring a new design.

The data table isn't used in a 100% correct way in the first place because each data type (name, description, added by, status) should go in a separate column. In its first implementation in https://github.com/WordPress/gutenberg/pull/36379, it was a `<ul>` list, styled like a table. It was then changed to a `<table>` element in https://github.com/WordPress/gutenberg/pull/36707 but at that point the table cells contained data of mixed type. The template 'source' and customized info were added in https://github.com/WordPress/gutenberg/pull/36763. As the cell data are a bit 'mixed', I'd tend to think using a table adds little value from an a11y perspective.

Aside: worth reminding this isn't the only place in the Editor where a dropdown menu with only one item is used. I'll create a new 'discussion' issue mentioning this PR

**Comment by afercia** [comment]
Re: the dropdown issue, see discussion at https://github.com/WordPress/gutenberg/discussions/42666

**Comment by paaljoachim** [comment]
Adding in this associated issue: 
[Site Editor] Templates list screen - Adding duplicate to 3 dot menu.
https://github.com/WordPress/gutenberg/issues/36665

**Comment by afercia** [comment]
Looking back into this, these action buttons are still barely visible.
@jameskoster @javierarce I appreciate your insights and feedback. When you have a chance, it would be great to find a solution for this issue. The simpler, the better 🙂 🙏

**Review by aristath — APPROVED** [review]
The code looks good and makes sense.
The new interface reduces cognitive load for me so personally I'd consider this one ready to be merged and a great improvement.
This looks OK to me to be merged so I'll go ahead and approve it but we'll also need some feedback from @WordPress/gutenberg-design before merging it or taking any further action 👍

**Comment by jameskoster** [comment]
I don't have a strong opinion either way, so happy to defer to other designers. I suspect we'll need to revisit the design for actions when we add bulk editing. 

Some minor points:

The stroke beneath each row in the table has vanished. It would be nice to reinstate that.

Padding on the button looks a bit awkward to me:

Screenshot 2023-04-12 at 12 08 35

The space on the right appears much larger than the left. Not sure if that's something to 'fix' here, or on the component itself. 

Is the `–` in empty action cells necessary? It seems like the lack of a button could be adequate?

**Comment by jasmussen** [comment]
Agreed, the "Actions" button is not a pattern we use anywhere else, whereas the ellipsis menu is. I recognize the intent of this issue and don't disagree that we need improvement, but I don't think this is the right way to go about it.

**Comment by aristath** [comment]
> Agreed, the "Actions" button is not a pattern we use anywhere else, whereas the ellipsis menu is. I recognize the intent of this issue and don't disagree that we need improvement, but I don't think this is the right way to go about it.

I agree, the "Actions" menu is not something we use anywhere. But perhaps we should? If not "Actions", then something else. What we currently have is beautiful for people with perfect vision who can see 3 tiny dots on the screen, but for everyone else it's less than ideal (myself included and I'm just a bit myopic like 30+% of the world).

@jasmussen Any ideas on alternative approaches for this? Suggestions on what we can use to improve the current UI?

**Comment by jarekmorawski** [comment]
Hi there! A WooCommerce designer here. I understand the concerns about the accessibility of the ellipsis button. No solution is perfect, but the way I see it, sticking to the ellipsis can be beneficial in the long term. The templates screen will likely set the foundations for other tables across wp-admin, WooCommerce, and other plugins—many of which are similar to or more complex than WordPress.

Extensions like WooCommerce need space to show the information the merchants need to make business decisions. We have multiple data-heavy tables, like products, orders, transactions, and subscriptions.

The ellipsis menu takes up little space and lets us display more information without forcing the user to scroll horizontally (or hide columns). It also lets us reduce complexity by hiding less common actions when they are not needed (for example, you can perform ~25 actions on each WooCommerce product).

This information is crucial, so the extra space we get by not using a regular button greatly helps. Here's how the products list in WooCommerce could look if the Templates list grew into becoming its own component shared across the platform (note that it's a WIP mockup; products come from MUJI).

Products

I've explored using the `Actions` button instead. With up to 100 products shown on the page simultaneously, the cognitive strain caused by added complexity and repeated information seemed just as significant.

Perhaps we could mitigate some accessibility issues by making the whole item row interactive? Clicking it would take you to the editor, so you would only need to use the ellipsis to perform a specific, less-common action.

**Comment by afercia** [comment]
> sticking to the ellipsis can be beneficial in the long term

I'd kindly disagree. It really depends on the way an ellipsis button is used and in what context.

## Conceptually
Ellipsis buttons are commonly used in many software. It's a common pattern to logically indicate 'More' options, settings, whatever. The key point here is the concept of 'More', which inherently assumes _there is something before_ the 'more'. In fact, in all the major software I can think of,  ellipsis 'more' button are used _toghether with_ other UI controls and usually _at the end_ of a series of controls. That's what 'more' means: a series of controls, buttons and the like, and then a 'more' with the less commonly used controls.

based on the above, by its own nature an ellipsis button should only be used within a series of other controls.

Instead, in the editor is now used everywhere. Often in isolation, as in: there's only the ellipsis button and no other controls before. This usage goes against the concept of 'more'. I'd argue that when used in isolation an ellipsis button is no longer a 'more', instead it's something else. As such, it brings in a different logical concept and it should not be represented by an ellipsis icon.

## Visually
When used in isolation, literally in the middle of a white space with no other controls or shapes around it, the editor ellipsis icon is barely visible for users with vision impairments. I do have vision impairments and I can't see this button. It's just three thin dots in the middle of a white space, I can't perceive these three thin dots and as such I really strugle to use this button.

It's important to understand the context matters a lot. Consider the 4 different cases illustrated in the screenshot below:

ellipsis button in context

1. Ellipsis 'Options' button at the right side of the editor top bar.
2. Ellipsis 'Options' button within a block toolbar.
3. Ellipsis 'Options' button within the block toolbar with the 'Top toolbar' mode enabled. Note that the top toolbar has now changed and it renders in a single row, but this is still a valid example for our purposes.
4. Ellipsis 'Actions' button in the Templates list.

For users with vision impairments, these 4 cases are very different. The ability to perceive the ellipsis icon also depends on the context as in: surrounding controls or borders or other shapes that help with visually perceiving the button.

Based on my (admittedly anectodical) experience:

1. There are other buttons before the the ellipsis. This helps 'driving' my vision and helps perceiving the eellipsis dots but still isn't that great.
2. The block toolbar buttons and more importantly the visible borders greatly help. I would say this is the case where I can better perceive the ellipsis.
3. Although there are buttons and borders, the color of the borders is a lighter gray. I struggle to perceive the ellipsis.
4. This is the worts case. The ellipsis button is literally placed in the middle of a white space with no other controls or borders around it. Also, it is on the extreme right side, which doesn't help. I really can't see this button unless I get at a very short distance from my screen. This usage is seriously problematic for any user with vision impairment and it's just a bad practice.

## Ellispis Proliferation 
I'd also like to highlight that the ellipsis button is now broadly used across the editor, too broadly in my opinion. There are scenarios when multiple ellipsis buttons are visible at the some moment in the user interface. Consider the screenshot below:

many ellipsis buttons at the same time

Up to 5 (maybe more) ellipsis buttons can be visible at the same time. I'm not sure this contributes in any wway to a good user experience. It's just confusing.

## Semantiics / Predictability 
An icon button should visually represent what it does. The underlying action / tool should always be the same, so that users understand on the fly what the button is meant for. The concept of 'more' would fit with this need. However, in the editor the ellipsis button is actually used for the most varied purposes. As said earlier, it's no longer a 'more'. it's something else and I'd  argue the inconsistent usage is any good for users. A few examples of the actions / features the ellipsis button is used for based on its labeling:

- Options (top bar)
- Styles actions
- Options for Post template
- Options for Columns
- Options (block toolbar)
- Actions (in the templates list)

**Comment by afercia** [comment]
I'm going to close this PR for two reasons:
- The templates list is going to be re-implemented with Data Views. As an aside, the Data Views also include a 'grid' layout as it was mentioned in the above comments. See all work on the Data Views here: https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Data%20Views
- The ellipsis button is still a problem when used in isolation. However, this is now a pattern that is used in several places in the editor. As such, it's now a much broader issue that warrants a broader discussion. I'll create a new issue.

**Pull request closed without merging**
